Serialize stats

sax opened the pull request:
This is the next bit of work for #177.

sax commented:
@rainliu This depends on the recent addition to `webrtc-ice`. I can update remove the change to pull that from github if/when that's released to crates.io.

Please let me know if you see any potential problems with these changes. I also ran clippy and found a few fixes to be made in stats generation.

k0nserv reviewed (commented):
```rust
        let lock = Arc::try_unwrap(collector).unwrap();
```

This line feels dangerous. How are we guaranteeing that there's only a single strong reference at this point(if there isn't this will panic)?

sax commented:
```rust
let lock = Arc::try_unwrap(collector).unwrap();
```

My thought was that if there is ever a weak reference to the collector open at this point, there is a bug somewhere else in the code that should be fixed. Just after this is the following:

```rust
let collector = lock.into_inner();

StatsReport {
  reports: collector.reports,
}
```

This needs to take ownership of the collector's reports.

I guess an alternative would be for the caller to unwrap the `Arc`, and for this to change from `impl From<Arc<Mutex<StatsCollector>>> for StatsReport` to `impl From<StatsCollector> for StatsReport`. Maybe that would make it less likely for this to introduce a bug later?

sax commented on src/stats/serialize.rs:
```diff
@@ -0,0 +1,21 @@
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards");
+
+        let epoch_ms = epoch.as_millis() as f64 / 1000.0;
```
This could definitely use a test, and perhaps some clarification of the code. The output of this is an approximate timestamp as a `f64`, with seconds to the left of the decimal point and milliseconds to the right.

sax replied:
In trying to write a test, the fact that the data sources are all `Instant`s makes things pretty difficult. There doesn't seem to be any way in the language to convert between instants and real time, and the docs go out of their way to say that instants cannot be compared to time.

The underlying data collected in stats is all `Instant`s, which is why I modeled stats that way. A longer term refactor might be to find the locations where those are being saved and convert them to `SystemTime`s or another data structure more closely connected to real world time... something where a specific instant in time can be hard coded into a test, and be serialized to a specific float.

k0nserv replied:
If we are only measuring elapsed time then `Instant` are exactly the right choice, that's what they are for after all. 

> This could definitely use a test, and perhaps some clarification of the code. The output of this is an approximate timestamp as a f64, with seconds to the left of the decimal point and milliseconds to the right.

Can we clarify this with a comment or rename the variable `epoch_seconds`?

sax replied:
That sounds great. Just pushed a change to rename to `instant_to_epoch_seconds`.

sax commented:
I think I'll need to go back to the Rust guides on weak references.

My understanding is that this `WaitGroup` should block until all usages are finished. Since all implementations of `collect_stats` take a wait group worker along with the `<Arc<Mutex<StatsCollectr>>`, each reference to the collector should be freed at the same time that the worker is freed.

sax commented:
This was just published: https://www.w3.org/TR/webrtc-stats/.

One thing I'm seeing is that the W3C recommendation has key names as lowerCamelCase, whereas I have been serializing to snake_case. I don't actually know what Pion does, but it might make sense to go towards the specifications.

sax commented:
Ok, I just switched it so that converting the stats collector to a stats report can bubble up an error if any strong references remain. I think I'm still missing something about weak references, however.

Arc mentions that `try_unwrap` will succeed if there are any weak references: https://doc.rust-lang.org/std/sync/struct.Arc.html#method.try_unwrap. The `WaitGroup` in PR should keep there from being any extraneous strong references, but just in case theres the `Result` that the caller can handle.

I would think that if anybody updated the code to `Arc::downgrade(collector)` somewhere, they would be responsible for making sure it had not been dropped if/when they called `weak_collector.upgrade`. https://doc.rust-lang.org/std/rc/struct.Weak.html#method.upgrade.

sax commented:
Notes from Discord:

> 1. Like we discussed before you can use Weak in PeerConnectionInternal::get_stats when passing the collector to each method. Related question, do we need Arc at all, can't we just use &StatsCollector?
> 1. I see a lot of try_unlock().unwrap() usage on the Mutex which will most certainly cause crashes. This should be just lock().unwrap() instead. 
> 1. I think it would be sensible to move the locking into StatsCollector itself and then it can expose methods that take &self such that the callers don't have to care  or know about the lock.
> 1. I see that we spawn a task in RTCIceGatherer::collect_stats, it think there's an argument for doing this at the very top level in PeerConnectionInternal::get_stats instead.
> 1. Do we really need the waitgroup? tokio:join already ensures that all the futures complete before returning and if we move the task spawning to the top level I don't think we need it?

I had added the wait group logic when starting this only because that's what Pion does, and I was a lot less familiar with tokio before working on these features. You're right that the `tokio::join!` waits for all branches to complete, so I've removed the wait groups.

As a part of that, I removed the spawned task from the `ice_gatherer`. I wonder if there's any actual benefit from spawning threads, or if it's fast enough to grab the statistics... everything will have to go through the mutex anyways.

I also changed the locks to use `lock().await`, and that passes the tests, so yay!

sax commented:
Getting rid of the thread actually seems to get rid of the need for the Arc, or for Weak. I think that's what I was getting hung up on.

Even if a thread is re-introduced, like you said it can be added at `PeerConnectionInternal::get_stats`, and the collector can live inside that thread without an Arc.

k0nserv commented:
I don't see a case where we would have exclusive access to `StatsCollector` i.e. it's always wrapped in `Mutex` when we use it. Based on this I think it would make sense to move the lock into the collector and use methods with `&self` receivers instead. Further, if the lock isn't held across `await` points it could be a `std::sync::Mutex` instead of a tokio Mutex.

WDYT @sax?

sax commented:
Now that the Arc and the thread are gone, I think this makes more sense to me. I finally have time today and next week to continue working on this, as well.

I think this has been a case where my inexperience with Rust led me to implement everything too similar to Pion, making everything more complicated than it needed to be.

sax commented:
Moving the lock into StatsCollector also makes it feel better to change it to a std Mutex. I started making that change recently, but with the unlocking distributed through the codebase it meant aliasing multiple versions of Mutex everywhere.

k0nserv commented:
That's great, glad we reached a nice state. Is this done now or is there more you need to get to before we are ready to merge?

k0nserv commented:
I think as long as these additions are forward compatible we are okay to release without them for now. Also would be good to try this out and see if there are any issues
